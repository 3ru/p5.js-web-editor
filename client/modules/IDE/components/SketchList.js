import React, { PropTypes } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import moment from 'moment';
import { Link } from 'react-router';
import * as SketchActions from '../actions/projects';
import * as ProjectActions from '../actions/project';
import InlineSVG from 'react-inlinesvg';
const exitUrl = require('../../../images/exit.svg');

class SketchList extends React.Component {
  componentDidMount() {
    this.props.getProjects(this.props.username);
  }

  render() {
    return (
      <div className="sketch-list">
        <header className="sketch-list__header">
          <h2>Sketches</h2>
          <button className="sketch-list__exit-button" onClick={this.props.closeSketchList}>
            <InlineSVG src={exitUrl} alt="Close Sketch List Overlay" />
          </button>
        </header>
        <div className="sketches-table-container">
          <table className="sketches-table" summary="table containing all saved projects">
            <thead>
              <tr>
                <th scope="col">Name</th>
                <th scope="col">Created</th>
                <th scope="col">Last Updated</th>
              </tr>
            </thead>
            <tbody>
              {this.props.sketches.map(sketch =>
                <tr className="sketches-table__row" key={sketch.id}>
                  <td scope="row"><Link to={`/projects/${sketch._id}`}>{sketch.name}</Link></td>
                  <td>{moment(sketch.createdAt).format('MMM D, YYYY h:mm:ss A')}</td>
                  <td>{moment(sketch.updatedAt).format('MMM D, YYYY h:mm:ss A')}</td>
                </tr>
              )}
            </tbody>
          </table>
        </div>
      </div>
    );
  }
}

SketchList.propTypes = {
  user: PropTypes.object.isRequired,
  getProjects: PropTypes.func.isRequired,
  sketches: PropTypes.array.isRequired,
  closeSketchList: PropTypes.func.isRequired,
  username: PropTypes.string
};

function mapStateToProps(state) {
  return {
    user: state.user,
    sketches: state.sketches
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(Object.assign({}, SketchActions, ProjectActions), dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(SketchList);
