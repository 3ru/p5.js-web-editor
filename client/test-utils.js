/**
 * This file re-exports @testing-library but ensures that
 * any calls to render have translations available.
 *
 * This means tested components will be able to call
 * `t()` and have the translations of the default
 * language
 *
 * See: https://react.i18next.com/misc/testing#testing-without-stubbing
 */

// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
import React from 'react';
import PropTypes from 'prop-types';
import { createStore } from 'redux';
import { Provider } from 'react-redux';

import { I18nextProvider } from 'react-i18next';
import i18n from './i18n-test';
import rootReducer from './reducers';

// re-export everything
// eslint-disable-next-line import/no-extraneous-dependencies
export * from '@testing-library/react';

const Providers = ({ children }) => (
  // eslint-disable-next-line react/jsx-filename-extension
  <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
);

Providers.propTypes = {
  children: PropTypes.element.isRequired
};

function reduxRender(
  ui,
  {
    initialState,
    store = createStore(rootReducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <I18nextProvider i18n={i18n}>
        <Provider store={store}>{children}</Provider>
      </I18nextProvider>
    );
  }

  Wrapper.propTypes = {
    children: PropTypes.element.isRequired
  };

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

const customRender = (ui, options) =>
  render(ui, { wrapper: Providers, ...options });

// override render method
export { customRender as render, reduxRender };
