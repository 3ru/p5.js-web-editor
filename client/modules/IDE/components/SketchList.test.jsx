import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import axios from 'axios';
import { unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import SketchList from './SketchList';
import { reduxRender, fireEvent, screen } from '../../../test-utils';
import {
  initialTestState,
  mockProjects
} from '../../../redux_test_stores/test_store';

jest.mock('../../../i18n');

/*
 * there seem to be polarizing opinions about whether or not
 * we should test the unconnected component or the
 * connected one. For the sake of not editing the original SketchList file
 * with an unneccessary export statement, I'm testing
 * the connected component with redux-mock-store.
 * this approach is outlined here -
 * https://www.robinwieruch.de/react-connected-component-test
 */

describe('<Sketchlist />', () => {
  let store;
  let container;
  const mockStore = configureStore([thunk]);

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
    axios.get.mockImplementationOnce((x) => Promise.resolve({ data: 'foo' }));
    store = mockStore(initialTestState);
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
    store.clearActions();
  });

  it('has sample projects', () => {
    let component;
    act(() => {
      component = reduxRender(<SketchList username="happydog1" />, {
        store,
        container
      });
    });
    expect(screen.getByText('testsketch1')).toBeInTheDocument();
    expect(screen.getByText('testsketch2')).toBeInTheDocument();
  });

  it('clicking on date created row header dispatches a reordering action', () => {
    let component;
    act(() => {
      component = reduxRender(<SketchList username="happydog2" />, {
        store,
        container
      });
    });
    act(() => {
      fireEvent.click(screen.getByTestId('toggle-direction-createdAt'));
    });
    const expectedAction = [{ type: 'TOGGLE_DIRECTION', field: 'createdAt' }];
    expect(store.getActions()).toEqual(expect.arrayContaining(expectedAction));
  });

  it('clicking on dropdown arrow opens sketch options', () => {
    let component;
    act(() => {
      component = reduxRender(<SketchList username="happydog2" />, {
        store,
        container
      });
    });
    const dropdown = screen.queryAllByTestId(
      'sketch-list-toggle-options-arrow'
    );

    if (dropdown.length > 0) {
      act(() => {
        fireEvent.click(dropdown[0]);
      });

      expect(screen.queryByText('Rename')).not.toBeInTheDocument();
      expect(screen.queryByText('Duplicate')).toBeInTheDocument();
      expect(screen.queryByText('Download')).toBeInTheDocument();
      expect(screen.queryByText('Add to collection')).toBeInTheDocument();
      expect(screen.queryByText('Delete')).not.toBeInTheDocument();
    }
  });

  it('clicking on dropdown arrow opens sketch options - sketches belong to user', () => {
    let component;
    act(() => {
      component = reduxRender(<SketchList username="happydog" />, {
        store,
        container
      });
    });
    const dropdown = screen.queryAllByTestId(
      'sketch-list-toggle-options-arrow'
    );

    if (dropdown.length > 0) {
      act(() => {
        fireEvent.click(dropdown[0]);
      });

      expect(screen.queryByText('Rename')).toBeInTheDocument();
      expect(screen.queryByText('Duplicate')).toBeInTheDocument();
      expect(screen.queryByText('Download')).toBeInTheDocument();
      expect(screen.queryByText('Add to collection')).toBeInTheDocument();
      expect(screen.queryByText('Delete')).toBeInTheDocument();
    }
  });

  it('snapshot testing', () => {
    const { asFragment } = reduxRender(<SketchList username="happydog2" />, {
      store,
      container
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
