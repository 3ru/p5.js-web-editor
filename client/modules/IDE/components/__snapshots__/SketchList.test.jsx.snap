// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sketchlist /> snapshot testing 1`] = `
<DocumentFragment>
  <article
    class="sketches-table-container"
  >
    <table
      class="sketches-table"
      summary="table containing all saved projects"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            <button
              aria-label="Sort by Sketch descending."
              class="sketch-list__sort-button"
              data-testid="toggle-direction-name"
            >
              <span
                class="sketches-table__header"
              >
                Sketch
              </span>
            </button>
          </th>
          <th
            scope="col"
          >
            <button
              aria-label="Sort by Date Created ascending."
              class="sketch-list__sort-button"
              data-testid="toggle-direction-createdAt"
            >
              <span
                class="sketches-table__header sketches-table__header--selected"
              >
                Date Created
              </span>
              <test-file-stub
                aria-label="Descending"
                focusable="false"
                role="img"
              />
            </button>
          </th>
          <th
            scope="col"
          >
            <button
              aria-label="Sort by Date Updated descending."
              class="sketch-list__sort-button"
              data-testid="toggle-direction-updatedAt"
            >
              <span
                class="sketches-table__header"
              >
                Date Updated
              </span>
            </button>
          </th>
          <th
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr
          class="sketches-table__row"
        >
          <th
            scope="row"
          >
            <a>
              testsketch1
            </a>
          </th>
          <td>
            Feb 25, 2021, 11:58:14 PM
          </td>
          <td>
            Feb 25, 2021, 11:58:29 PM
          </td>
          <td
            class="sketch-list__dropdown-column"
          >
            <button
              aria-label="Toggle Open/Close Sketch Options"
              class="sketch-list__dropdown-button"
              data-testid="sketch-list-toggle-options-arrow"
            >
              <test-file-stub
                aria-hidden="true"
                focusable="false"
              />
            </button>
          </td>
        </tr>
        <tr
          class="sketches-table__row"
        >
          <th
            scope="row"
          >
            <a>
              testsketch2
            </a>
          </th>
          <td>
            Feb 23, 2021, 12:40:43 PM
          </td>
          <td>
            Feb 23, 2021, 12:40:43 PM
          </td>
          <td
            class="sketch-list__dropdown-column"
          >
            <button
              aria-label="Toggle Open/Close Sketch Options"
              class="sketch-list__dropdown-button"
              data-testid="sketch-list-toggle-options-arrow"
            >
              <test-file-stub
                aria-hidden="true"
                focusable="false"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </article>
</DocumentFragment>
`;
